rdkPlugins/Gamepad/source/GamepadPlugin.cpp

```diff
@@ -0,0 +1,176 @@
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include "GamepadPlugin.h"
+
+// #include <linux/limits.h>
+// #include <sys/mount.h>
+// #include <sys/stat.h>
+// #include <sys/types.h>
+// #include <fcntl.h>
+// #include <unistd.h>
+// #include <mntent.h>
+#include <iostream>
+#include <sstream>
+
+REGISTER_RDK_PLUGIN(GamepadPlugin);
+
+GamepadPlugin::GamepadPlugin(std::shared_ptr<rt_dobby_schema> &containerConfig,
+                     const std::shared_ptr<DobbyRdkPluginUtils> &utils,
+                     const std::string &rootfsPath)
+    : mName("Gamepad"),
+      mContainerConfig(containerConfig),
+      mUtils(utils)
+{
+    AI_LOG_FN_ENTRY();
+    AI_LOG_FN_EXIT();
+}
+
+unsigned GamepadPlugin::hookHints() const
+{
+    return (IDobbyRdkPlugin::HintFlags::PostInstallationFlag);
+}
+
+// Begin Hook Methods
+
+/**
+ * @brief Dobby Hook - run in host namespace *once* when container bundle is downloaded
+ */
+bool GamepadPlugin::postInstallation()
+{
+    AI_LOG_FN_ENTRY();
+
+    // 1. add devices
+    rt_config_linux_resources* schemaResources = mContainerConfig->linux->resources;
+    size_t old_devices_len = schemaResources->devices_len;
+    const int FIRST_CONTROLLER = 64;
+    const int NUM_CONTROLLERS = 10;
+
+    schemaResources->devices_len += NUM_CONTROLLERS;
+    schemaResources->devices = (rt_defs_linux_device_cgroup**)realloc(schemaResources->devices, sizeof(rt_defs_linux_device_cgroup*) * schemaResources->devices_len);
+
+    rt_defs_linux_device_cgroup* device = schemaResources->devices[old_devices_len];
+
+    for (int i = 0; i < NUM_CONTROLLERS; ++i)
+    {
+        device = (rt_defs_linux_device_cgroup*)calloc(1, sizeof(rt_defs_linux_device_cgroup));
+        device->type = strdup("c");
+        device->access = strdup("rw");
+        device->major = 13;
+        device->major_present = true;
+        device->minor = FIRST_CONTROLLER + i;
+        device->minor_present = true;
+        device->allow = true;
+        device->allow_present = true;
+
+        schemaResources->devices[old_devices_len + i] = device;
+        device++;
+    }
+
+
+    // 2. add /dev/input mount, "nodev"(MS_NODEV) flag removed
+    mUtils->addMount("/dev/input/", "/dev/input/", "bind", {"bind", "nosuid", "noexec"});
+
+
+    // 3. add input group to gidMapping
+    int inputGroupId = getInputGroupId();
+    //printf("inputGroupId: %d\n", inputGroupId);
+    rt_defs_id_mapping* newMapping = (rt_defs_id_mapping*)calloc(1, sizeof(rt_defs_id_mapping));
```

Comment:
Please create add_group_id (or similar) function and use it to separate the logic

Reply:
Done